### DISCUSSION TITLE
Speedscope-like format

### DISCUSSION THREAD

**Discussion by pnf** [async-profiler/async-profiler#815] [Ideas]
When this has come up before, @apangin, you emphasized that jfr was the recommended default format and seemed to imply that at some point that would be the only format. 
I want to argue for the value of of a format that
- summarizes the profiling period rather than reporting every event
- is therefore more compact than jfr
- is accessible in-process without recourse to the file system (like collapsed stacks today)
- is normalized, at least to the extent of tokening frame names
- supports multiple, simultaneously profiles events

The speedscope [0.6 json format](https://github.com/jlfwong/speedscope/blob/main/sample/profiles/speedscope/0.6.0/two-sampled.speedscope.json) satisfies these criteria - it's pretty much the obvious json format one would imagine. 

In our build, already modified for other reasons, we do something quite a bit hackier, which I obviously don't recommend:
- "etypeframes" => add a frame for the type of event
- "memoframes" => memoize frame names with Dictionary, reporting them in collapsed format first as 123=mymethodname and subsequently as 123.

**Comment by aalexand** [comment]
Related: https://github.com/open-telemetry/oteps/pull/237

**Comment by apangin** [comment]
Thank you for opening the discussion. While I agree with most of the points raised, the Speedscope format does not seem a good fit to me.
1. JSON format is neither human readable nor computer friendly. It's excessive and expensive.
2. The structure is not compact: each stack trace includes a separate copy of all frames (in contrast to a tree-like structure).
3. It is not aware of Java specifics like class names. It does not have a notion of frame type (although this can be easily added as a Frame attribute).
4. There is just one "weight" per frame that does not allow expressing easily states like runnable/sleeping, interpreted/compiled/inlined or TLAB vs. outside TLAB allocations.

OTel proposal actually suffers from the same issues, except that protobuf is slightly better in terms of size and efficiency.

**Reply by nevgeniev** [reply]
do you think that dumping stack traces in sbe (simple binary encoding) to agrona (broadcaster) ring buffer in /dev/shm is a bad idea? it seems to me like a win/win combination - portable/fast  and gives you a stream of stack traces w/o disk io + should be easy to implement.. :) having public sbe schema will allow anyone to implement own "processor" of such shm queue
plus you can put all unique frames in indexed cache in the same shm.. so actual events will be very compact (just index of the cached string)

**Reply by apangin** [reply]
Do you propose a solution to the aforementioned problem or just suggest another format to support?
I think the original question was about a format for aggregated profiling data.

**Reply by nevgeniev** [reply]
yep, sorry I've misread original qn

**Comment by pnf** [comment]
I agree that speedscope is far from ideal, which is why I went for more obviously temporary hack to solve our immediate problems. A tree representation (probably an adjacency list) in particular would be valuable (though even less readable by a human).  A richer weight/label for leaf nodes is easy to envisage.
Right now, the choice in officially released async-profiler is between collapsed stacks, which is extraordinarly limited, and jfr, which is much too expensive to use for real-time observability across many processes.

**Comment by parttimenerd** [comment]
You can still use a converter like [jfrtofp](https://github.com/parttimenerd/jfrtofp) to convert a JFR file to speedscope format:

```
java -jar jfrtofp-all.jar samples/small_profile.jfr --mode speedscope
```

But I have to agree with @apangin that the speedscope format is not made to represent Java profiling information.

**Comment by pnf** [comment]
Converting from jfr is possible, but it’s a lot of overhead when one doesn’t actually need every event, especially when aggregating over 100s or 1000s of processes.   In fact, pyroscope falls over trying to incorporate even a few dozen jfrs in parallel.
Again, I am not arguing that speedscope json is an ideal or even a particularly good format, just that there is a need for a compact representation that is aggregated over a time.  



> 
> 
> You can still use a converter like jfrtofp <https://github.com/parttimenerd/jfrtofp> to convert a JFR file to speedscope format:
> 
> `java -jar jfrtofp-all.jar samples/small_profile.jfr --mode speedscope
`
